crates/driver/src/domain/competition/order/mod.rs

```diff
@@ -329,3 +381,71 @@ impl Jit {
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn order_scaling() {
+        let sell = |amount: u64| eth::Asset {
+            token: eth::H160::from_low_u64_be(0x5e11).into(),
+            amount: eth::U256::from(amount).into(),
+        };
+        let buy = |amount: u64| eth::Asset {
+            token: eth::H160::from_low_u64_be(0xbbbb).into(),
+            amount: eth::U256::from(amount).into(),
+        };
+        let weth = eth::WethAddress(eth::H160([0xef; 20]).into());
+
+        let order = |sell_amount: u64, buy_amount: u64, executed: Option<eth::Asset>| Order {
+            uid: Default::default(),
+            receiver: Default::default(),
+            valid_to: util::Timestamp(u32::MAX),
+            buy: buy(buy_amount),
+            sell: sell(sell_amount),
+            side: match executed {
+                None => Side::Sell,
+                Some(executed) if executed.token == sell(0).token => Side::Sell,
+                Some(executed) if executed.token == buy(0).token => Side::Buy,
+                _ => panic!(),
+            },
+            fee: Default::default(),
+            kind: Kind::Limit,
+            app_data: Default::default(),
+            partial: executed
+                .map(|executed| Partial::Yes {
+                    executed: executed.amount.into(),
+                })
+                .unwrap_or(Partial::No),
+            pre_interactions: Default::default(),
+            post_interactions: Default::default(),
+            sell_token_balance: SellTokenBalance::Erc20,
+            buy_token_balance: BuyTokenBalance::Erc20,
+            signature: Signature {
+                scheme: signature::Scheme::PreSign,
+                data: Default::default(),
+                signer: Default::default(),
+            },
+        };
+
+        assert_eq!(order(1000, 1000, Some(sell(250))).solver_sell(), sell(750));
+        assert_eq!(order(1000, 1000, Some(sell(250))).solver_buy(weth), buy(750));
+        assert_eq!(order(1000, 1000, Some(buy(250))).solver_sell(), sell(750));
+        assert_eq!(order(1000, 1000, Some(buy(250))).solver_buy(weth), buy(750));
+
+        assert_eq!(order(1000, 100, Some(sell(99))).solver_sell(), sell(901));
+        assert_eq!(order(1000, 100, Some(sell(99))).solver_buy(weth), buy(91));
+
+        assert_eq!(order(100, 1000, Some(buy(99))).solver_sell(), sell(90));
+        assert_eq!(order(100, 1000, Some(buy(99))).solver_buy(weth), buy(901));
+
+        assert_eq!(order(1000, 1, Some(sell(500))).solver_sell(), sell(500));
+        assert_eq!(order(1000, 1, Some(sell(500))).solver_buy(weth), buy(1));
+
+        assert_eq!(order(1, 1000, Some(buy(500))).solver_sell(), sell(0));
```

Comment:
This example is a bit contrived since you can't really partially fill an order with 1wei sell amount (needs to be at least two). One test case that I think should pass (but doesn't at the moment). Would be:

```
assert_eq!(order(2, 1000, Some(buy(501))).solver_sell(), sell(1));
```

Reply:
Hmm, why do you think that should pass? AFAIU, that should round to a "sell 0 for 499" order.

1/499 does not respect the 2/1000 limit price.